test/fail_compilation/test17423.d

```diff
@@ -0,0 +1,29 @@
+/* REQUIRED_ARGS: -dip1000
+TEST_OUTPUT:
+---
+fail_compilation/test17423.d(26): Error: reference to local `this` assigned to non-scope parameter `dlg` calling test17423.Bar.opApply
+---
+*/
+
+// https://issues.dlang.org/show_bug.cgi?id=17423
+
+struct Bar
+{
+    int delegate(int) @safe myDlg;
+
+    int opApply(int delegate(int) @safe dlg) @safe {
+        myDlg = dlg;
+        return 0;
+    }
+}
+
+struct Foo
+{
+    Bar o;
+    int i = 3;
+
+    this(int x) @safe {
+        foreach(_; o) { i = 0; }
+        i = x;
+    }
+}
```

Comment:
The test file from the bug report still triggers in a segfault (with this PR).

Reply:
It fails to compile with `-dip1000`

Reply:
Oh my point was that 17423 isn't fixed by this PR because the segfault still exists and with `-dip1000` it currently doesn't segfault: https://run.dlang.io/is/W1DGlg

Reply:
My point is -dip1000 is required to fix it. The whole point of dip1000 is to fix these sorts of issues.